bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/BusinessPartnerSearchService.kt

```diff
@@ -0,0 +1,216 @@
+    private val logger = KotlinLogging.logger { }
+
+    /**
+     * Uses the [searchRequest] and [paginationRequest] to perform query on database for business partner.
+     * The BPNs of found partners are used to query the whole business partner records from the database.
+     * The records are supplied with relevancy scores of the search hits and returned as a paginated result.
+     * In case BPNs, can not be found in the database, the [PageDto] properties are
+     * adapted accordingly from the page information
+     *
+     */
+    fun searchLegalEntities(
+        searchRequest: BusinessPartnerSearchRequest,
+        paginationRequest: PaginationRequest
+    ): PageDto<LegalEntityMatchVerboseDto> {
+        val legalEntityPage = searchAndPrepareLegalEntityPage(searchRequest, paginationRequest)
+        businessPartnerFetchService.fetchLegalEntityDependencies(legalEntityPage.content.map { (_, legalEntity) -> legalEntity }.toSet())
+
+        return with(legalEntityPage) {
+            PageDto(totalElements, totalPages, page, contentSize,
+                content.map { (score, legalEntity) -> legalEntity.toMatchDto(score) })
+        }
+    }
+
+    private fun searchAndPrepareLegalEntityPage(
+        searchRequest: BusinessPartnerSearchRequest,
+        paginationRequest: PaginationRequest
+    ): PageDto<Pair<Float, LegalEntity>> {
+        return if (searchRequest == BusinessPartnerSearchRequest.EmptySearchRequest) {
+            paginateLegalEntities(paginationRequest)
+        } else {
+            searchLegalEntity(searchRequest, paginationRequest)
+        }
+    }
+
+    private fun paginateLegalEntities(paginationRequest: PaginationRequest): PageDto<Pair<Float, LegalEntity>> {
+        logger.debug { "Paginate database for legal entities" }
+        val legalEntityPage = legalEntityRepository.findAll(PageRequest.of(paginationRequest.page, paginationRequest.size))
+
+        return legalEntityPage.toDto(legalEntityPage.content.map { Pair(0f, it) }) // assign 0 score as no search has been conducted
+    }
+
+    private fun searchLegalEntity(
+        searchRequest: BusinessPartnerSearchRequest,
+        paginationRequest: PaginationRequest
+    ): PageDto<Pair<Float, LegalEntity>> {
+        logger.debug { "Search index for legal entities" }
+        val value = searchRequest.partnerProperties.legalName.toString()
+        val legalEntityPage = legalEntityRepository.findByLegalNameValue(value, PageRequest.of(paginationRequest.page, paginationRequest.size))
+        return PageDto(
+            totalElements = legalEntityPage.totalElements,
+            totalPages = legalEntityPage.totalPages,
+            page = paginationRequest.page,
+            content = legalEntityPage.content.map { Pair(legalEntityPage.totalElements - levenshteinDistance(it.legalName.value, value), it) },
+            contentSize = legalEntityPage.content.size
+        )
+
+    }
+
+    private fun levenshteinDistance(text1: String, text2: String): Float {
```

Comment:
Retroactively applying levenstein leads to possible faulty ordering. Either the distance is applied on database level or we work without it. I would remove it and work with length based ordering for now, like suggested in the comment above

Reply:
Done as mentioned in above comment.